keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,577 @@
+
+### Proxy gRPC definition
+
+```grpc
+syntax = "proto3";
+
+service ProxyService {
+    // Proxy connects to a remote address by stream id, and establish
+    // a bi-directional stream of packet exchange.
+    rpc Proxy(stream Packet) returns (stream Packet) {}
+}
+
+enum PacketType {
+  DIAL_REQ = 0;
+  DIAL_RSP = 1;
+  CLOSE_REQ = 2;
+  CLOSE_RSP = 3;
+  DATA = 4;
+}
+
+message Packet {
+  PacketType type = 1;
+
+  oneof payload {
+    DialRequest dialRequest = 2;
+    DialResponse dialResponse = 3;
+    CloseRequest closeRequest = 4;
+    CloseResponse closeResponse = 5;
+    Data data = 6;
+  }
+}
+
+// Error is sent when error happens from remote side of connection when it tries
+// to read or write from it
+enum Error {
+  EOF = 0;
+  EIO = 1;
+  ECONNRESET = 2;
+  ETIMEOUT = 3;
+  EADDRNOTAVAIL = 4;
+  EMFILE = 5;
+  // … 
+}
+
+// DialRequest represents a request to dial to an address on the other 
+// side of the tunnel. The format is inspired by golang's net interface
+// https://golang.org/pkg/net/#Dial 
+message DialRequest {
+    // network representing a named network. "Tcp" is the only supported
+    // value
+    string network = 1;
+
+    // For TCP network, the address has the form "ip:port", where host must
+    // be IP address.
+    string address = 2;
+
+    // random is the randomly generated bytes that represents a given dial request.
+    // the number is kept unchanged across the proxies, and copied to DialResponse
+    // from the other side of the tunnel. The number cannot be reused across 
+    // different DialRequests.
+    bytes random = 3;
+}
+
+// DialResponse is the response to a DialRequest.
+message DialResponse {
+    // Error when the dial request cannot be fulfilled.
+    Error error = 1;
+
+    // connectID represents a unique connect ID for the connection/stream.
+    int64 connectID = 2;
+
+    // Copied from DialRequest. Explained also in DialRequest.
+    bytes random = 3;
+}
+
+// CloseRequest requests to close a connection
+message CloseRequest {
+    // connection id to close.
+    int64 connectID = 1;
+}
+
+// CloseResponse is the response to a close connection request.
+message CloseResponse {
+    // connection id to close.
+    int64 connectionID = 1;
+
+    // Error when close request cannot be fulfilled.
+    Error error = 2;
+}
+
+message Data {
+    // connectID that data or error payload is belonged to
+    int32 connectID = 1;
+
+    // data payload
+    bytes data = 2;
+
+    // Error that happens to the connection.
+    Error error = 3;
+}
+```
+
+### Connectivity Proxy
+
+The connectivity proxy(s) can run in the same container as the KAS. 
+It should run on the same machine and must run in the same flat network as the KAS. 
+It listens on a port for gRPC connections from the KAS. 
+This port would be for forwarding traffic to the appropriate cluster. 
+It should have an administrative port speaking https. 
+The administrative port serves the liveness probe and metrics. 
```

Comment:
Good point. My thought was that as this is a third port (master side being the first and node/cluster/agent side being the second, both of which should have mTLS) that its restrictions could be different. I was also thinking that it would probably only be listening on localhost. If you feel that it still needs mTLS I can seperate the probe ports from the rest of the admin ports.